artifactory/services/groups.go

```diff
@@ -0,0 +1,89 @@
+package services
+
+import (
+	"encoding/json"
+	"fmt"
+	rthttpclient "github.com/jfrog/jfrog-client-go/artifactory/httpclient"
+	"github.com/jfrog/jfrog-client-go/auth"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+)
+
+type Group struct {
+	Name            string `json:"name,omitempty"`
+	Description     string `json:"description,omitempty"`
+	AutoJoin        bool   `json:"autoJoin,omitempty"`
+	AdminPrivileges bool   `json:"adminPrivileges,omitempty"`
+	Realm           string `json:"realm,omitempty"`
+	RealmAttributes string `json:"realmAttributes,omitempty"`
+}
+type GroupService struct {
+	client     *rthttpclient.ArtifactoryHttpClient
+	ArtDetails auth.ServiceDetails
+}
+
+func NewGroupService(client *rthttpclient.ArtifactoryHttpClient) *GroupService {
+	return &GroupService{client: client}
+}
+func (gs *GroupService) SetArtifactoryDetails(rt auth.ServiceDetails) {
+	gs.ArtDetails = rt
+}
+
+func (gs *GroupService) GetGroup(name string) (*Group, error) {
+	httpDetails := gs.ArtDetails.CreateHttpClientDetails()
+	url := fmt.Sprintf("%sapi/security/groups/%s", gs.ArtDetails.GetUrl(), name)
+	res, body, _, err := gs.client.SendGet(url, true, &httpDetails)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode > 204 {
+		return nil, fmt.Errorf("%d %s: %s", res.StatusCode, res.Status, string(body))
+	}
+	var group Group
+	if err := json.Unmarshal(body, &group); err != nil {
+		return nil, errorutils.CheckError(err)
+	}
+	return &group, nil
+}
+
+func (gs *GroupService) CreateOrUpdateGroup(group Group) error {
+	httpDetails := gs.ArtDetails.CreateHttpClientDetails()
+	content, err := json.Marshal(group)
+	if err != nil {
+		return err
+	}
+
+	httpDetails.Headers = map[string]string{
+		"Content-Type": "application/json",
+		"Accept":       "application/json",
+	}
+	//
+	url := fmt.Sprintf("%sapi/security/groups/%s", gs.ArtDetails.GetUrl(), group.Name)
+	resp, body, err := gs.client.SendPut(url, content, &httpDetails)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode > 204 {
```

Comment:
See my comments in the closed PR - it is prefered not to check ranges of HTTP response codes. Let's check that the expected value is returned, and return an error otherwise.

Reply:
Eyal,

I had no desire to do this - except that when I was getting 400 errors `err` was equal to `nil` - I couldn't use err as an indicator. And, the low level client code doesn't consider 400 an error. I will revisit this code.

Reply:
If you look at this, code, you basically realize that if there is no `err`, only then check the statusCode. 

From what I can tell, the go socket code won't return `err` unless it's like a `SO_TIMEOUT`
See this [sof](https://stackoverflow.com/questions/55210301/error-handling-with-http-newrequest-in-go)
Here are the [official docs](https://golang.org/pkg/net/http/)

I am really forced to check the status code

Reply:
<img width="880" alt="Screen Shot 2020-12-30 at 10 23 19 AM" src="https://user-images.githubusercontent.com/1476816/103373556-1bf64a00-4a8a-11eb-8a7b-a44c2a63ef65.png">

You have examples in code by your team that do this very thing

Reply:
Here is the [code in question](https://github.com/jfrog/jfrog-client-go/blob/dev/artifactory/services/buildinfo.go#L78)

Reply:
Can you please replace code like
```
	if resp.StatusCode > 204 {
		return fmt.Errorf("%d %s", resp.StatusCode, resp.Status)
	}
```
with:
```
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("%d %s", resp.StatusCode, resp.Status)
	}
```
This is what I meant when suggested not to check respose code ranges. Anything other than what you expect to receive as a successful response code, is an error.

Reply:
Sure, I think I can accommodate.